### PR TITLE
Add serializers and API for forms and questions

### LINKED ISSUE #26: As an Admin, I am able to create forms and add questions on backend using API
## Description
As an Admin, I am able to create forms and questions using API
## Mocks
N/A

## Acceptance Criteria
- [ ] Create viewsets for form feedback
- [ ] Create serializers for form feedback
- [ ] Crate endpoints for form feedback
- [ ] Write tests
- [ ] Add checks for form feedback

## Definition of Done
- [ ] All of the required items are completed.
- [ ] Approval by 1 mentor.

## Estimation
3-4 days

### PR DESCRIPTION
**Pull request by bismitaguha**
### Description

- Add serializer for form, questions and inherited models

Fixes #26 

### Type of Change:

- Code

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

N/A

### Checklist:

- [x] My PR follows the style guidelines of this project
- [ ] I have performed a self-review of my own code or materials
- [x] Any dependent changes have been merged

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been published in downstream modules

### CODE REVIEW

**Comment by bismitaguha** [comment]
@Monal5031 I need to discuss about the editing fields feature. So I will join your office hours! That's left for this PR.

File: osp/serializers/fields.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+from osp.models import (
+    Question, Choice, Checkbox, Dropdown,
+    Paragraph, ShortAnswer, Date, Time, FileUpload
+)
+from osp.serializers.question import QuestionSerializer
+
+class ChoiceSerializer(serializers.ModelSerializer):
+
+    class Meta(QuestionSerializer):
```
Comment (Monal5031) [inline osp/serializers/fields.py]: You don't need to inherit `QuestionSerializer` here but rather in `ChoiceSerializer`

File: osp/serializers/fields.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
+from rest_framework import serializers
+
+from osp.models import (
+    Question, Choice, Checkbox, Dropdown,
+    Paragraph, ShortAnswer, Date, Time, FileUpload
+)
+from osp.serializers.question import QuestionSerializer
+
+class ChoiceSerializer(serializers.ModelSerializer):
```
Comment (Monal5031) [inline osp/serializers/fields.py]: I don't see some explicit logic for creation of options?
Is it working properly without explicit logic?

File: osp/serializers/form.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from rest_framework import serializers
+
+from osp.models import Form
+
+class FormSerializer(serializers.ModelSerializer):
+
+    class Meta:
+        model = Form
+        fields = [
+            'id',
```
Comment (Monal5031) [inline osp/serializers/form.py]: 🙄 condensation...

File: osp/serializers/question.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+from rest_framework import serializers
+
+from osp.models import Question
+
+class QuestionSerializer(serializers.ModelSerializer):
+
+    class Meta:
+
+        model = Question
+        fields = [
```
Comment (Monal5031) [inline osp/serializers/question.py:10]: same here

File: osp/views/form.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        if self.request.method == 'GET':
+            self.permission_classes = [IsAuthenticated]
+        else:
+            self.permission_classes = [IsAuthenticated, IsAdmin]
+
+        return super(FormView, self).get_permissions()
+
+    def get_queryset(self):
+        status = self.request.query_params.get('status', None)
+        queryset = Form.objects.filter(published_status=status).order_by('-created_on')
```
Comment (Monal5031) [inline osp/views/form.py]: queryset should be defined on class level and used as: `queryset = self.queryset`

File: osp/views/question.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            self.permission_classes = [IsAuthenticated, IsAdmin]
+
+        return super(QuestionView, self).get_permissions()
+
+    
+    # GET request
+    def list(self, request):
+        form_id = self.request.query_params.get('form_id', None)
+        fields = Form.objects.filter(id=form_id).values_list('form_fields', flat=True)
+        objs = Question.objects.filter(id__in=fields).order_by('order')
```
Comment (Monal5031) [inline osp/views/question.py]: we can define ordering on the meta class of model?

File: osp/views/question.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        for obj in objs:
+            value = get_model_and_serializer(obj.data_type)
+            model = value['model']
+            serializer = value['serializer']
+            instance = model.objects.get(id=obj.id)
+            results.append(serializer(instance).data)
+            
+        #response
+        return Response(results)
```
Comment (Monal5031) [inline osp/views/question.py]: status code

File: osp/views/question.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        else:
+            self.permission_classes = [IsAuthenticated, IsAdmin]
+
+        return super(QuestionView, self).get_permissions()
+
+    
+    # GET request
+    def list(self, request):
+        form_id = self.request.query_params.get('form_id', None)
+        fields = Form.objects.filter(id=form_id).values_list('form_fields', flat=True)
```
Comment (Monal5031) [inline osp/views/question.py]: why do we need this? Each question has form as FK so form_id can directly be used right?

File: osp/views/question.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            instance = model.objects.get(id=obj.id)
+            results.append(serializer(instance).data)
+            
+        #response
+        return Response(results)
+
+
+    # POST request
+    def create(self, request):
+        form_id = self.request.query_params.get('form_id', None)
```
Comment (Monal5031) [inline osp/views/question.py]: query params are used for sorting/fetching purposes, if you need to create/update data, it needs to be sent in the body

Comment (Monal5031) [reply osp/views/question.py]: Also if you send it inside body of each question obj then I think line 69 `form.form_fields.add()` would also become redundant.

File: osp/views/question.py [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+            value = get_model_and_serializer(obj['data_type'])
+            model = value['model']
+            serializer = value['serializer']
+            id = obj.get('id', None)
+            if id is not None:
+                model.objects.filter(id=id).update(**obj)
+                instance = model.objects.get(id=id)
+            else:
+                instance = model.objects.create(**obj)
+            print(instance)
```
Comment (Monal5031) [inline osp/views/question.py]: remove print before commiting

**Review by Monal5031 — CHANGES_REQUESTED** [review]

File: osp/views/question.py (hunk 8 above)
Comment (bismitaguha) [reply osp/views/question.py]: form_id is not foreign key, it's in many to many

**Review by Monal5031 — APPROVED** [review]

**Pull request merged**